lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,512 @@
+                throw new Error(
+                    "MultiFactorAuth recipe has already been initialised. Please check your code for bugs."
+                );
+            }
+        };
+    }
+
+    static reset() {
+        if (process.env.TEST_MODE !== "testing") {
+            throw new Error("calling testing function in non testing env");
+        }
+        Recipe.instance = undefined;
+    }
+
+    // abstract instance functions below...............
+
+    getAPIsHandled = (): APIHandled[] => {
+        return [
+            {
+                method: "get",
+                pathWithoutApiBasePath: new NormalisedURLPath(GET_MFA_INFO),
+                id: GET_MFA_INFO,
+                disabled: this.apiImpl.mfaInfoGET === undefined,
+            },
+        ];
+    };
+
+    handleAPIRequest = async (
+        id: string,
+        _tenantId: string,
+        req: BaseRequest,
+        res: BaseResponse,
+        _: NormalisedURLPath,
+        __: HTTPMethod,
+        userContext: UserContext
+    ): Promise<boolean> => {
+        let options = {
+            recipeInstance: this,
+            recipeImplementation: this.recipeInterfaceImpl,
+            config: this.config,
+            recipeId: this.getRecipeId(),
+            isInServerlessEnv: this.isInServerlessEnv,
+            req,
+            res,
+        };
+        if (id === GET_MFA_INFO) {
+            return await mfaInfoAPI(this.apiImpl, options, userContext);
+        }
+        throw new Error("should never come here");
+    };
+
+    handleError = async (err: STError, _: BaseRequest, __: BaseResponse): Promise<void> => {
+        throw err;
+    };
+
+    getAllCORSHeaders = (): string[] => {
+        return [];
+    };
+
+    isErrorFromThisRecipe = (err: any): err is STError => {
+        return STError.isErrorFromSuperTokens(err) && err.fromRecipe === Recipe.RECIPE_ID;
+    };
+
+    addGetAllFactorsFromOtherRecipesFunc = (f: GetAllFactorsFromOtherRecipesFunc) => {
+        this.getAllFactorsFromOtherRecipesFunc.push(f);
+    };
+
+    getAllAvailableFactorIds = (tenantConfig: TenantConfig) => {
+        let factorIds: string[] = [];
+        for (const func of this.getAllFactorsFromOtherRecipesFunc) {
+            const factorIdsRes = func(tenantConfig);
+            factorIds = factorIds.concat(factorIdsRes.factorIds);
+        }
+        return factorIds;
+    };
+
+    getAllAvailableFirstFactorIds = (tenantConfig: TenantConfig) => {
+        let factorIds: string[] = [];
+        for (const func of this.getAllFactorsFromOtherRecipesFunc) {
+            const factorIdsRes = func(tenantConfig);
+            factorIds = factorIds.concat(factorIdsRes.firstFactorIds);
+        }
+        return factorIds;
+    };
+
+    addGetFactorsSetupForUserFromOtherRecipes = (func: GetFactorsSetupForUserFromOtherRecipesFunc) => {
+        this.getFactorsSetupForUserFromOtherRecipesFuncs.push(func);
+    };
+
+    validateForMultifactorAuthBeforeFactorCompletion = async ({
```

Comment:
This isn't exposed to devs and in most cases I don't think this would be useful as it is mostly concerned with: 
- calling `isAllowedToSetupFactor`, which is exposed separately
- the way we link users and the related security considerations, which doesn't really apply to custom factors since they are not using ST users